lib/ui/recommendation.js

```diff
@@ -187,5 +221,164 @@ Recommendation.prototype = {
+    }
+  },
+  duplicateCheck: function() {
+    // Given recommendation and results data, check if the recommendation
+    // duplicates one of the results:
+    // if it's a wikipedia recommendation, a duplicate has the exact same URL;
+    // if it's a domain recommendation, a duplicate is any page on that domain.
+
+    // Don't check for duplicates until we have both pieces of data.
+    if (!this.searchResults || !this.recData) {
+      return;
+    }
+
+    let isDuplicate = false;
+    let type = 'tld' in this.recData.enhancements ? 'tld' : 'wikipedia';
+
+    // nsIURI gives us a quick and easy way to normalize URLs, reducing
+    // comparison failures due to minor variations. Start by constructing
+    // an nsIURI object from the recommendation URL. If the input was
+    // malformed, bail.
+    let recommendationUri;
+    try {
+      recommendationUri = this.io.newURI(this.recData.result.url, null, null);
+    } catch (ex) {
+      return;
+    }
+
+    // Loop over the results, create a URI object for each, then compare the
+    // URI objects. If a duplicate is found, we know not to show the
+    // recommendation, so bail immediately.
+    for (let i = 0; i < this.searchResults.length; i++) {
+      let result = this.searchResults[i];
+      let resultUri;
+
+      // Sometimes the top heuristic result will just be a bare domain, lacking
+      // the scheme ("http" or "https") and '://' part. The nsIURI constructor
+      // will throw without the scheme, so we must prepend either 'http://' or
+      // 'https://'. Wikipedia uses https, and we do exact matching on
+      // wikipedia recommendations when checking for duplicates, while for TLD
+      // recommendations, we compare base domains, ignoring the scheme.
+      // Therefore, we prepend 'https://' if `result` lacks a scheme.
+      //
+      // Finally, note that this check ignores moz-action URLs, which might
+      // have a URL in a serialized JSON object later in the string, but we
+      // want to deal with moz-action results separately.
+      if (result.indexOf('http') !== 0) {
+        result = 'https://' + result;
+      }
+
+      try {
+        resultUri = this.io.newURI(result, null, null);
+      } catch (ex) {
+        continue;
+      }
+
+      // If the result is a moz-action URL, we might need to either replace
+      // the resultUri, if the moz-action contains a URL inside it; or bail
+      // on the current result, if the moz-action doesn't contain a URL.
+      if (this._isMozAction(resultUri)) {
+        resultUri = this._replaceMozAction(resultUri);
+        if (!resultUri) {
+          continue;
+        }
+      }
+
+      // Finally, check for duplicates.
+      // For TLD results, the eTLD service should give us a good, cheap guess
+      // at the domain.
+      // For wikipedia results, use nsIURI.equals() to check for URL equality.
+      // If a duplicate is found, immediately exit; else, show the
+      // recommendation.
```

Comment:
I think this code is excessively documented. Maybe just document surprising / weird things (like the moz-action stuff). There's no need to say in English exactly what the code says in JavaScript.